### PR TITLE
feat: add deleteitem function to write interface

### PR DESCRIPTION
**Pull request by vankongv**
Added for DeleteItem to WriteAPI interface.
Updated README for correct docker run command

### CODE REVIEW

**Review by shidil — APPROVED** [review]

**Comment by shidil** [comment]
Prefer table driven tests. They allow adding new cases easily.

**Review by shidil — CHANGES_REQUESTED** [review]
Test does not check if it deletes more items than expected.
Try to add more seed data to the table than just one.
When there are more items, verify if the correct item was removed.

File: tests/delete_item_test.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		err := table.PutItem(ctx, dynago.StringValue(item.Pk), dynago.StringValue(item.Sk), item)
+		if err != nil {
+			t.Fatalf("failed to insert item %v; got %s", item, err)
+		}
+	}
+
+	//query all items to check they exist
+	for _, item := range items {
+		var output []user
+		_, err := table.Query(ctx, "pk = :pk and sk = :sk", map[string]types.AttributeValue{
```
Comment (shidil) [inline tests/delete_item_test.go]: Use GetItem or perform a single query here. It doesn't really matter much as far as output is concerned. However, it shows lack of understanding or concern about DynamoDB

Comment (vankongv) [reply tests/delete_item_test.go]: image
Updated with GetItem for fetching single items

File: tests/delete_item_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
+package tests
```
Comment (NalinSajwan) [inline tests/delete_item_test.go]: As discussed I have added the sample code for your test cases (which allows multiple cases).

> Seed records which can be used `tests/seedTable.go`:
```go
package tests

import (
	"context"
	"fmt"

	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
	"github.com/oolio-group/dynago"
)

type tableRecord struct {
	Pk     string `dynamodbav:"pk"`
	Sk     string `dynamodbav:"sk"`
	Record interface{}
}

func seedRecords(ctx context.Context, table *dynago.Client, input []tableRecord) error {
	items := make([]*dynago.TransactPutItemsInput, 0, len(input))
	for _, item := range input {
		in := dynago.TransactPutItemsInput{
			PartitionKeyValue: dynago.StringValue(item.Pk),
			SortKeyValue:      dynago.StringValue(item.Sk),
			Item:              item,
		}

		items = append(items, &in)
	}

	err := table.TransactPutItems(ctx, items)
	if err != nil {
		return fmt.Errorf("failed to insert items; got %s", err)
	}

	requests := make([]map[string]types.AttributeValue, 0, len(input))
	for _, item := range input {
		req := map[string]types.AttributeValue{
			"pk": dynago.StringValue(item.Pk),
			"sk": dynago.StringValue(item.Sk),
		}
		requests = append(requests, req)
	}

	var output []tableRecord
	err = table.BatchGetItems(ctx, requests, &output)
	if err != nil {
		return fmt.Errorf("failed to get items; got %s", err)
	}

	if len(input) != len(output) {
		return fmt.Errorf("count of items does not match input items; got %s", err)
	}

	return nil
}
```

> For `tests/delete_item_test.go`:
```go
package tests

import (
	"context"
	"fmt"
	"slices"
	"testing"

	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
	"github.com/oolio-group/dynago"
)

type user struct {
	Name  string
	Phone string
}

type userKeys struct {
	pk string
	sk string
}

func getSeedData(ctx context.Context, table *dynago.Client, seedData []tableRecord) ([]tableRecord, error) {
	itemsToGet := make([]map[string]types.AttributeValue, 0, len(seedData))
	for _, v := range seedData {
		itemsToGet = append(itemsToGet, map[string]types.AttributeValue{
			"pk": dynago.StringValue(v.Pk),
			"sk": dynago.StringValue(v.Sk),
		})
	}

	var out []tableRecord

	err := table.BatchGetItems(ctx, itemsToGet, &out)
	if err != nil {
		return nil, fmt.Errorf("Unable to get seed data: %w", err)
	}

	return out, nil
}

func TestDeleteItem(t *testing.T) {
	ctx := context.TODO()
	table := prepareTable(t, dynamoEndpoint, "delete_test")

	records := []tableRecord{
		{
			Pk: "users#org1",
			Sk: "user#1",
			Record: user{
				Name:  "User 1",
				Phone: "xyz",
			},
		},
		{
			Pk: "users#org1",
			Sk: "user#2",
			Record: user{
				Name:  "User 2",
				Phone: "xyz",
			},
		},
		{
			Pk: "users#org1",
			Sk: "user#3",
			Record: user{
				Name:  "User 3",
				Phone: "xyz",
			},
		},
	}

	type testCase struct {
		title         string
		itemsToDelete []userKeys
		itemsLeft     int
		seedData      []tableRecord
	}

	cases := []testCase{
		{
			title: "Delete User 2",
			itemsToDelete: []userKeys{
				{
					pk: "users#org1",
					sk: "user#2",
				},
			},
			itemsLeft: 2,
			seedData:  records,
		},
		{
			title: "Delete all users",
			itemsToDelete: []userKeys{
				{
					pk: "users#org1",
					sk: "user#1",
				},
				{
					pk: "users#org1",
					sk: "user#2",
				},
				{
					pk: "users#org1",
					sk: "user#3",
				},
			},
			itemsLeft: 0,
			seedData:  records,
		},
		{
			title: "Delete wrong users",
			itemsToDelete: []userKeys{
				{
					pk: "users#org1",
					sk: "user#none",
				},
			},
			itemsLeft: 3,
			seedData:  records,
		},
	}

	for _, c := range cases {
		t.Run(c.title, func(t *testing.T) {
			err := seedRecords(ctx, table, c.seedData)
			if err != nil {
				t.Fatalf("failed to prepare records; got %s", err)
			}

			for _, item := range c.itemsToDelete {
				err = table.DeleteItem(ctx, dynago.StringValue(item.pk), dynago.StringValue(item.sk))
				if err != nil {
					t.Fatalf("unable to delete records; got %s", err)
				}
			}

			data, err := getSeedData(ctx, table, c.seedData)
			if err != nil {
				t.Fatalf("failed to prepare records; got %s", err)
			}

			// see if "itemsToDelete" are the ones deleted.
			dataKeys := make([]string, 0, len(data))
			for _, item := range data {
				dataKeys = append(dataKeys, fmt.Sprintf("%s--%s", item.Pk, item.Sk))
			}

			for _, v := range c.itemsToDelete {
				recKey := fmt.Sprintf("%s--%s", v.pk, v.sk)
				if slices.Contains(dataKeys, recKey) {
					t.Fatalf("expected items to deleted in db but found it, pk %s; sk %s", v.pk, v.sk)
				}
			}

			// compare items remaining is seed data matches test
			if len(data) != c.itemsLeft {
				t.Fatalf("expected items in db; %v found; %v", c.itemsLeft, len(data))
			}
		})
	}
}
```

**Review by NalinSajwan — COMMENTED** [review]
Added code enhancing test cases.

**Comment by vankongv** [comment]
image
New test cases passed

File: tests/delete_item_test.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+	var out []tableRecord
+
+	err := table.BatchGetItems(ctx, itemsToGet, &out)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to get seed data: %w", err)
+	}
+
+	return out, nil
+}
```
Comment (NalinSajwan) [inline tests/delete_item_test.go]: This actually part of seed table, I forgot to move it there. 

```suggestion
```

File: tests/seedTable.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+			Item:              item,
+		}
+
+		items = append(items, &in)
+	}
+
+	err := table.TransactPutItems(ctx, items)
+	if err != nil {
+		return fmt.Errorf("failed to insert items; got %s", err)
+	}
```
Comment (NalinSajwan) [inline tests/seedTable.go]: This can be moved to a function `putSeedData` and called by `seedRecords`.

File: tests/seedTable.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+			"sk": dynago.StringValue(item.Sk),
+		}
+		requests = append(requests, req)
+	}
+
+	var output []tableRecord
+	err = table.BatchGetItems(ctx, requests, &output)
+	if err != nil {
+		return fmt.Errorf("failed to get items; got %s", err)
+	}
```
Comment (NalinSajwan) [inline tests/seedTable.go]: Can be a function of its own `getSeedData` and called in `seedRecord` to verify everything got added.

File: delete_item.go [hunk 6]
```diff
@@ -14,20 +14,20 @@ import (
 * @param sk the sort key of the record
  * @return true if the record was deleted, false otherwise
 */
-func (t *Client) DeleteItem(ctx context.Context, pk string, sk string) error {
+func (t *Client) DeleteItem(ctx context.Context, pk Attribute, sk Attribute) error {
 
 	//delete item from dynamodb
```
Comment (shidil) [inline delete_item.go]:
```suggestion
```

File: delete_item.go [hunk 7]
```diff
@@ -14,20 +14,20 @@ import (
...
 		TableName: &t.TableName,
 		Key: map[string]types.AttributeValue{
-			"pk": &types.AttributeValueMemberS{Value: pk},
-			"sk": &types.AttributeValueMemberS{Value: sk},
+			"pk": pk,
+			"sk": sk,
 		},
 	}
-	resp, err := t.client.DeleteItem(ctx, input)
+	_, err := t.client.DeleteItem(ctx, input)
```
Comment (shidil) [inline delete_item.go]:
```suggestion
```

**Review by shidil — APPROVED** [review]

**Pull request merged**